File: java/src/main/java/net/netheos/pcsapi/oauth/PasswordSessionManager.java
/**
 * Copyright (c) 2014 Netheos (http://www.netheos.net)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.netheos.pcsapi.oauth;

import net.netheos.pcsapi.credentials.PasswordCredentials;
import net.netheos.pcsapi.credentials.UserCredentialsRepository;
import net.netheos.pcsapi.exceptions.CRetriableException;
import net.netheos.pcsapi.request.CResponse;
import net.netheos.pcsapi.storage.StorageBuilder;
import net.netheos.pcsapi.utils.PcsUtils;

import org.apache.http.HttpResponse;
import org.apache.http.auth.AuthScope;
import org.apache.http.auth.UsernamePasswordCredentials;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpUriRequest;
import org.apache.http.client.protocol.ClientContext;
import org.apache.http.protocol.HttpContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.net.URI;
import java.util.HashMap;
import java.util.Map;
import org.apache.http.HttpHost;
import org.apache.http.client.CredentialsProvider;
import org.apache.http.impl.client.BasicCredentialsProvider;
import org.apache.http.protocol.BasicHttpContext;

/**
 * Password session manager : handles basic and digest authentication.
 */
public class PasswordSessionManager
        extends SessionManager<PasswordCredentials>
{

    private static final Logger LOGGER = LoggerFactory.getLogger( PasswordSessionManager.class );

    private final UserCredentialsRepository userCredentialsRepo;
    private final HttpClient httpClient;
    private final UsernamePasswordCredentials usernamePasswordCredentials;
    /**
     * Cache of HttpContext, required to do preemptive authentications.
     *
     * These objects should not be shared between threads, so they are created and wrapped in a ThreadLocal.
     */
    private final Map<HttpHost, ThreadLocal<HttpContext>> cache = new HashMap<HttpHost, ThreadLocal<HttpContext>>();

    public PasswordSessionManager( StorageBuilder builder )
    {
        super( builder.getUserCredentials() );

        this.httpClient = builder.getHttpClient();
        this.userCredentialsRepo = builder.getUserCredentialsRepo();

        usernamePasswordCredentials = new UsernamePasswordCredentials( userCredentials.getUserId(),
                                                                       userCredentials.getCredentials().getPassword() );

        // checks if we already have user_credentials:
        if ( userCredentials != null ) {
            PasswordCredentials credentials = userCredentials.getCredentials();
            if ( credentials.getPassword() == null ) {
                throw new IllegalStateException( "User credentials do not contain any password" );
            }
        }
    }

    @Override
    public CResponse execute( HttpUriRequest request )
    {
        if ( LOGGER.isTraceEnabled() ) {
            LOGGER.trace( "{}: {}", request.getMethod(), PcsUtils.shortenUrl( request.getURI() ) );
        }

        try {
            URI uri = request.getURI();
            HttpHost host = new HttpHost( uri.getHost(), uri.getPort(), uri.getScheme() );
            HttpContext context = getHttpContext( host );

            HttpResponse httpResponse = httpClient.execute( host, request, context );
            return new CResponse( request, httpResponse );

        } catch ( IOException ex ) {
            // a low level error should be retried :
            throw new CRetriableException( ex );
        }
    }

    private synchronized HttpContext getHttpContext( HttpHost host )
    {
        ThreadLocal<HttpContext> tlContext = cache.get( host );
        if ( tlContext == null ) {
            tlContext = new ThreadLocal<HttpContext>();
            cache.put( host, tlContext );
        }
        HttpContext context = tlContext.get();
        if ( context == null ) {
            AuthScope scope = new AuthScope( host.getHostName(), host.getPort() );
            CredentialsProvider credentialsProvider = new BasicCredentialsProvider();
            credentialsProvider.setCredentials( scope, usernamePasswordCredentials );

            context = new BasicHttpContext();
            context.setAttribute( ClientContext.CREDS_PROVIDER, credentialsProvider );
            tlContext.set( context );
        }
        return context;
    }

    UserCredentialsRepository getUserCredentialsRepository()
    {
        return userCredentialsRepo;
    }

}

File: cpp/libpcs_api/src/auth/password_session_manager.cc
/**
 * Copyright (c) 2014 Netheos (http://www.netheos.net)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <mutex>

#include "cpprest/http_helpers.h"
#include "cpprest/json.h"
#include "cpprest/asyncrt_utils.h"

#include "pcs_api/password_credentials.h"
#include "pcs_api/c_exceptions.h"
#include "pcs_api/internal/password_session_manager.h"
#include "pcs_api/internal/uri_utils.h"
#include "pcs_api/internal/request_invoker.h"
#include "pcs_api/internal/logger.h"

namespace pcs_api {

PasswordSessionManager::PasswordSessionManager(const StorageBuilder& builder,
                                               const web::uri& base_uri)
    : clients_pool_(base_uri, builder.http_client_config()) {
    // Check type of credentials, if provided:
    std::shared_ptr<UserCredentials> p_user_creds =
                                                builder.GetUserCredentials();
    if (!p_user_creds) {
        BOOST_THROW_EXCEPTION(CStorageException(
                                "No user credentials provided"));
    }
    PasswordCredentials *p_password_credentials =
            dynamic_cast<PasswordCredentials *>(&p_user_creds->credentials());
    if (p_password_credentials == nullptr) {  // cast failed ?
        BOOST_THROW_EXCEPTION(CStorageException(
                "Invalid credentials type (expected PasswordCredentials)"));
    }
    web::credentials web_credentials(
        utility::conversions::to_string_t(p_user_creds->user_id()),
        utility::conversions::to_string_t(p_password_credentials->password()));
    // We add user credentials to the given client configuration
    // (this impacts pool):
    builder.http_client_config()->set_credentials(web_credentials);
}

std::shared_ptr<CResponse> PasswordSessionManager::Execute(
                                            web::http::http_request request) {
    LOG_TRACE << utility::conversions::to_utf8string(request.method())
              << ": "
              << UriUtils::ShortenUri(request.request_uri());

    // Take a client from the pool:
    std::unique_ptr<web::http::client::http_client> p_client =
        std::unique_ptr<web::http::client::http_client>(clients_pool_.Get());
    // will be copied when given to CResponse:
    pplx::cancellation_token_source cancel_source;
    web::http::http_response response = p_client->request(
                                            request,
                                            cancel_source.get_token()).get();
    // give client to response, and function to release client:
    auto p_response = std::make_shared<CResponse>(
            p_client.get(),
            std::bind(&PasswordSessionManager::ReleaseClient,
                      this,
                      std::placeholders::_1),
            request, &response, cancel_source);
    p_client.release();
    return p_response;
}


void PasswordSessionManager::ReleaseClient(
                                web::http::client::http_client *p_client) {
    // This is how we release clients: put them back in our pool:
    clients_pool_.Put(p_client);
}

}  // namespace pcs_api
